src/overlays/actors/ovl_En_River_Sound/z_en_river_sound.c

```diff
@@ -48,37 +49,57 @@ void EnRiverSound_Init(Actor* thisx, GlobalContext* globalCtx) {
 void EnRiverSound_Destroy(Actor* thisx, GlobalContext* globalCtx) {
     EnRiverSound* this = (EnRiverSound*)thisx;
 
-    if (this->actor.params == RS_SARIAS_SONG) {
+    if (this->actor.params == RS_LOST_WOODS_SARIAS_SONG) {
         Audio_ClearSariaBgmAtPos(&this->actor.projectedPos);
-    } else if (this->actor.params == RS_UNK_13) {
+    } else if (this->actor.params == RS_GORON_CITY_SARIAS_SONG) {
         Audio_ClearSariaBgm2();
     }
 }
 
-s32 func_80AE6A54(Vec3f* arg0, Vec3f* arg1, Vec3f* arg2, Vec3f* arg3) {
-    Vec3f vec[3];
+/**
+ * Determines the closest point to hearPos that is contained on the line segment connecting points A & B
+ * If the closest point on the line segment will be either end points A or B, return false
+ * Otherwise, calculate the point between A & B, assign it to `newSoundPos`, and return true
+ */
+s32 EnRiverSound_FindClosestPointOnLineSegment(Vec3f* pointA, Vec3f* pointB, Vec3f* hearPos, Vec3f* newSoundPos) {
+    Vec3f lineSeg[3];
     f32 temp;
 
-    vec[0].x = arg0->x - arg2->x;
-    vec[0].y = arg0->y - arg2->y;
-    vec[0].z = arg0->z - arg2->z;
-
-    vec[1].x = arg1->x - arg2->x;
-    vec[1].y = arg1->y - arg2->y;
-    vec[1].z = arg1->z - arg2->z;
-
-    vec[2].x = vec[1].x - vec[0].x;
-    vec[2].y = vec[1].y - vec[0].y;
-    vec[2].z = vec[1].z - vec[0].z;
-
-    temp = DOTXYZ(vec[2], vec[0]);
-
-    if ((DOTXYZ(vec[2], vec[1]) * temp) < 0.0f) {
-        temp = -temp / (SQ(vec[2].x) + SQ(vec[2].y) + SQ(vec[2].z));
-
-        arg3->x = (vec[2].x * temp) + arg0->x;
-        arg3->y = (vec[2].y * temp) + arg0->y;
-        arg3->z = (vec[2].z * temp) + arg0->z;
+    // Line Segment from the hearPos to the first point in the path
+    lineSeg[0].x = pointA->x - hearPos->x;
+    lineSeg[0].y = pointA->y - hearPos->y;
+    lineSeg[0].z = pointA->z - hearPos->z;
+
+    // Line Segment from the hearPos to the second point in the path
+    lineSeg[1].x = pointB->x - hearPos->x;
+    lineSeg[1].y = pointB->y - hearPos->y;
+    lineSeg[1].z = pointB->z - hearPos->z;
+
+    // Line Segment from the first point to the second point in the path
+    lineSeg[2].x = lineSeg[1].x - lineSeg[0].x;
+    lineSeg[2].y = lineSeg[1].y - lineSeg[0].y;
+    lineSeg[2].z = lineSeg[1].z - lineSeg[0].z;
+
+    temp = DOTXYZ(lineSeg[2], lineSeg[0]);
+
+    /**
+     *  |                   |
+     *  |                   |
+     *  |                   |
+     *  A ----------------- B
+     *  |                   |
+     *  |                   |
+     *  |                   |
+     * This condition uses dot products to check to see that `hearPos` is contained within the above region
+     * i.e. The closest point on line segment AB must be between A & B based on perpindicular distance
```

Comment:
```suggestion
     * i.e. The closest point on line segment AB must be between A & B based on perpendicular distance
```
Also I think "line segment" should just be "line"

I don't think "based on perpendicular distance" is even that useful, that's already covered by "closest point" and assuming euclidean geometry

Reply:
The reason I used the term "line segment" instead of line is to emphasize that the point on the line that's closest to `hearPos` must be on the section between A and B. If it falls on the line outside of these points, then that is the condition in which it returns false. The mathematical way to differentiate this is a "line segment" which connects two points, as opposed to a "line" which extends infinitely.

That said, that's based more on math instead of OoT's convention. What are your thoughts about keeping or dropping the "segment" part from that info?

I agree though that the perpendicular distance part is redundant. Also thanks for catching spelling mistakes :)

![image](https://user-images.githubusercontent.com/47598039/147885642-1a169387-dacd-49c6-973e-2464f72b3d5c.png)

Reply:
I think line segment/LineSeg is fine to use in the function name, but I don't understand why `Vec3f lineSeg[3];` is called lineSeg.

Reply:
The array in question is specifically referring to these three vectors that connect points `P`, `A`, and `B`. The term `vec` feels too general to me as literally anything that is `Vec3f` could be a vec. This seems to be three line segments that connect points PAB in a triangle. Although I'll admit the use of the word "line segment" is stretching it a bit, it still makes more sense to me than just vec.

Any thoughts for alternative names? It could also be named after a triangle in a sense too, since the three points do form a triangle shape. And one could interpret the check to return true/false as a check to see if the triangle is acute (all internal angles are less than 90 degrees). However, thinking of the problem in terms of triangles, while equivalent, is a more abstract way of thinking about it one step away from rivers and points.

Open to suggestions :)

![image](https://user-images.githubusercontent.com/47598039/147896209-c02c7b82-5848-460d-97c3-7913c0f2ee89.png)